README.md
# script.rss.editor

changelog.txt
4.0.1
- fix broken ok and yesno dialogs
- skin improvements

4.0.0
- update for python 3

3.0.4
- fix string problem

3.0.3
- update language directory structure
- update python except syntax
- add assets tag in addon.xml

3.0.2
- cosmetics

3.0.1
- fixes

3.0.0
- update skin to match Estuary
- fixes and cosmetics

2.0.4
- fix crash when no RssFeeds.xml file is found

2.0.3
- bump for Frodo

2.0.2
- fix encoding issues when writing to log

2.0.1
- fix encoding issues when opening files

2.0.0
- eden compatible version


lib/rssEditor.py
import os
import sys
import xbmc
import xbmcgui
import xbmcaddon
from .xmlParser import XMLParser

#enable localization
getLS = xbmcaddon.Addon().getLocalizedString
CWD = xbmcaddon.Addon().getAddonInfo('path')

class GUI(xbmcgui.WindowXMLDialog):
    def __init__(self, *args, **kwargs):
        xbmcgui.WindowXMLDialog.__init__(self, *args, **kwargs)
        self.setNum = kwargs['setNum']
        self.parser = XMLParser()
        if self.parser.feedsTree:
            self.doModal()

    def onInit(self):
        self.defineControls()
        self.feedsList = self.parser.feedsList[self.setNum]['feedslist'] #shortname
        if not self.feedsList:
            xbmcgui.Dialog().ok(getLS(40) + 'RssFeeds.xml', 'RssFeeds.xml ' + getLS(32041), getLS(32042), getLS(32043))
            self.closeDialog()
        self.showDialog()

    def defineControls(self):
        #actions
        self.action_cancel_dialog = ( 9, 10, 92, 216, 247, 257, 275, 61467, 61448 )
        #control ids
        self.control_heading_label_id = 2
        self.control_list_label_id = 4
        self.control_list_id = 10
        self.control_changeSet_button_id = 11
        self.control_add_button_id = 13
        self.control_remove_button_id = 14
        self.control_ok_button_id = 18
        self.control_cancel_button_id = 19
        #controls
        self.heading_label = self.getControl(self.control_heading_label_id)
        self.list_label = self.getControl(self.control_list_label_id)
        self.list = self.getControl(self.control_list_id)
        self.add_button = self.getControl(self.control_add_button_id)
        self.remove_button = self.getControl(self.control_remove_button_id)
        self.changeSet_button = self.getControl(self.control_changeSet_button_id)
        self.ok_button = self.getControl(self.control_ok_button_id)
        self.cancel_button = self.getControl(self.control_cancel_button_id)

    def showDialog(self):
        self.heading_label.setLabel(getLS(32000))
        self.list_label.setLabel(getLS(32012))
        self.changeSet_button.setLabel(getLS(32001))
        self.updateFeedsList()
        self.setFocus(self.list)

    def closeDialog(self):
        self.close()

    def onClick(self, controlId):
        #edit existing feed
        if controlId == self.control_list_id:
            position = self.list.getSelectedPosition()
            oldUrl = self.feedsList[position]['url']
            oldUpdateInterval = self.feedsList[position]['updateinterval']
            newUrl, newUpdateInterval = self.getNewFeed(oldUrl, oldUpdateInterval)
            if newUrl:
                self.feedsList[position] = {'url':newUrl, 'updateinterval':newUpdateInterval}
            self.updateFeedsList()
        #add new feed
        elif controlId == self.control_add_button_id:
            newUrl, newUpdateInterval = self.getNewFeed()
            if newUrl:
                self.feedsList.append({'url':newUrl, 'updateinterval':newUpdateInterval})
            self.updateFeedsList()
        #remove existing feed
        elif controlId == self.control_remove_button_id:
            self.removeFeed()
            self.updateFeedsList()
        #change/modify set
        elif controlId == self.control_changeSet_button_id:
            from . import setEditor
            setEditorUI = setEditor.GUI("script-RSS_Editor.xml", CWD, "default", setNum = self.setNum)
            self.close()
            del setEditorUI
        #save xml
        elif controlId == self.control_ok_button_id:
            self.parser.writeXmlToFile()
            self.closeDialog()
        #cancel dialog
        elif controlId == self.control_cancel_button_id:
            self.closeDialog()

    def onAction(self, action):
        if action in self.action_cancel_dialog:
            self.closeDialog()

    def onFocus(self, controlId):
        pass

    def removeFeed(self):
        position = self.list.getSelectedPosition()
        self.feedsList.remove(self.feedsList[position])
        #add empty feed if last one is deleted
        if len(self.feedsList) < 1:
            self.feedsList = [{'url':'http://', 'updateinterval':'30'}]

    def getNewFeed(self, url = 'http://', newUpdateInterval = '30'):
        kb = xbmc.Keyboard(url, getLS(32012), False)
        kb.doModal()
        if kb.isConfirmed():
            newUrl = kb.getText()
            newUpdateInterval = xbmcgui.Dialog().numeric(0, getLS(32013), newUpdateInterval)
        else:
            newUrl = None
        return newUrl, newUpdateInterval

    def updateFeedsList(self):
        self.list.reset()
        for feed in self.feedsList:
            self.list.addItem(feed['url'])
        if self.setNum == 'set1':
            self.list_label.setLabel(getLS(32014) % (''))
        else:
            self.list_label.setLabel(getLS(32014) % ('(' + self.setNum + ')'))


lib/xmlParser.py
import os
import sys
import unicodedata
from xml.dom.minidom import parse, Document, _write_data, Node, Element
import xbmc
import xbmcaddon
import xbmcgui

def log(txt):
    xbmc.log(msg=txt, level=xbmc.LOGDEBUG)

def writexml(self, writer, indent="", addindent="", newl=""):
    #credit: http://ronrothman.com/public/leftbraned/xml-dom-minidom-toprettyxml-and-silly-whitespace/
    writer.write(indent + "<" + self.tagName)
    attrs = self._get_attributes()
    a_names = list(attrs.keys())
    a_names.sort()
    for a_name in a_names:
        writer.write(" %s=\"" % a_name)
        _write_data(writer, attrs[a_name].value)
        writer.write("\"")
    if self.childNodes:
        if len(self.childNodes) == 1 \
          and self.childNodes[0].nodeType == Node.TEXT_NODE:
            writer.write(">")
            self.childNodes[0].writexml(writer, "", "", "")
            writer.write("</%s>%s" % (self.tagName, newl))
            return
        writer.write(">%s" % (newl))
        for node in self.childNodes:
            node.writexml(writer, indent + addindent, addindent, newl)
        writer.write("%s</%s>%s" % (indent, self.tagName, newl))
    else:
        writer.write("/>%s" % (newl))

# monkey patch to fix whitespace issues with toprettyxml
Element.writexml = writexml
#enable localization
getLS = xbmcaddon.Addon().getLocalizedString


class XMLParser:
    def __init__(self):
        self.RssFeedsPath = xbmc.translatePath('special://userdata/RssFeeds.xml')
        sane = self.checkRssFeedPathSanity()
        if sane:
            try:
                self.feedsTree = parse(self.RssFeedsPath)
            except:
                log('RSS Editor --> Failed to parse ' + unicodedata.normalize( 'NFKD', self.RssFeedsPath ))
                regen = xbmcgui.Dialog().yesno(getLS(32040), getLS(32051), getLS(32052), getLS(32053))
                if regen:
                    log('RSS Editor --> Attempting to Regenerate RssFeeds.xml')
                    xml = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n<rssfeeds>\n\
                    <!-- RSS feeds. To have multiple feeds, just add a feed to the set. You can also have multiple sets. 	!-->\n\
                    <!-- To use different sets in your skin, each must be called from skin with a unique id.             	!-->\n\
                    <set id="1">\n    <feed updateinterval="30">http://feeds.kodi.tv/xbmc</feed>\n    \
                    <feed updateinterval="30">http://feeds.kodi.tv/latest_xbmc_addons</feed>\n    \
                    <feed updateinterval="30">http://feeds.kodi.tv/updated_xbmc_addons</feed></set>\n</rssfeeds>'
                    f = open(self.RssFeedsPath, 'w')
                    f.write(xml)
                    f.close()
                    self.__init__()
                else:
                    log('RSS Editor --> User opted to not regenerate RssFeeds.xml.  Script Exiting')
                    self.feedsTree = False
            if self.feedsTree:
                self.feedsList = self.getCurrentRssFeeds()
        else:
            self.feedsTree = False
            self.feedsList = False
            log('RSS Editor --> Could not open ' + unicodedata.normalize( 'NFKD', self.RssFeedsPath ) + '. Either the file does not exist, or its size is zero.')

    def checkRssFeedPathSanity(self):
        if os.path.isfile(self.RssFeedsPath):
            #If the filesize is zero, the parsing will fail.  XBMC creates 
            if os.path.getsize(self.RssFeedsPath):
                return True

    def getCurrentRssFeeds(self):
        feedsList = dict()
        sets = self.feedsTree.getElementsByTagName('set')
        for s in sets:
            setName = 'set' + s.attributes["id"].value
            feedsList[setName] = {'feedslist':list(), 'attrs':dict()}
            #get attrs
            for attrib in s.attributes.keys():
                feedsList[setName]['attrs'][attrib] = s.attributes[attrib].value
            #get feedslist
            feeds = s.getElementsByTagName('feed')
            for feed in feeds:
                feedsList[setName]['feedslist'].append({'url':feed.firstChild.toxml(), 'updateinterval':feed.attributes['updateinterval'].value})
        return feedsList

    def formXml(self):
        """Form the XML to be written to RssFeeds.xml"""
        #create the document
        doc = Document()
        #create root element
        rssfeedsTag = doc.createElement('rssfeeds')
        doc.appendChild(rssfeedsTag)
        #create comments
        c1Tag = doc.createComment('RSS feeds. To have multiple feeds, just add a feed to the set. You can also have multiple sets.')
        c2Tag = doc.createComment('To use different sets in your skin, each must be called from skin with a unique id.')
        rssfeedsTag.appendChild(c1Tag)
        rssfeedsTag.appendChild(c2Tag)
        for setNum in sorted(self.feedsList.keys()):
            #create sets
            setTag = doc.createElement('set')
            #create attributes
            setTag.setAttribute('id', self.feedsList[setNum]['attrs']['id'])
            #only write rtl tags if they've been explicitly set
            if 'rtl' in self.feedsList[setNum]['attrs']:
                setTag.setAttribute('rtl', self.feedsList[setNum]['attrs']['rtl'])
            rssfeedsTag.appendChild(setTag)
            #create feed elements
            for feed in self.feedsList[setNum]['feedslist']:
                feedTag = doc.createElement('feed')
                feedTag.setAttribute('updateinterval', feed['updateinterval'])
                feedUrl = doc.createTextNode(feed['url'])
                feedTag.appendChild(feedUrl)
                setTag.appendChild(feedTag)
        return doc.toprettyxml(indent = '  ', encoding = 'UTF-8')

    def writeXmlToFile(self):
        log('RSS Editor --> writing to %s' % (unicodedata.normalize( 'NFKD', self.RssFeedsPath)))
        xml = self.formXml()
        #hack for standalone attribute, minidom doesn't support DOM3
        xmlHeaderEnd = xml.find(b'?>')
        xml = xml[:xmlHeaderEnd] + b' standalone="yes"' + xml[xmlHeaderEnd:]
        try:
            RssFeedsFile = open(self.RssFeedsPath, 'wb')
            RssFeedsFile.write(xml)
            RssFeedsFile.close()
            log('RSS Editor --> write success')
            self.refreshFeed()
        except IOError as error:
            log('RSS Editor --> write failed', error)

    def refreshFeed(self):
        """Refresh XBMC's rss feed so changes can be seen immediately"""
        xbmc.executebuiltin('refreshrss()')


lib/setEditor.py
import os
import sys
import xbmc
import xbmcgui
import xbmcaddon
from .xmlParser import XMLParser

#enable localization
getLS = xbmcaddon.Addon().getLocalizedString
CWD = xbmcaddon.Addon().getAddonInfo('path')

class GUI(xbmcgui.WindowXMLDialog):
    def __init__(self, *args, **kwargs):
        xbmcgui.WindowXMLDialog.__init__(self, *args, **kwargs)
        self.setNum = kwargs['setNum']
        self.parser = XMLParser()
        if self.parser.feedsTree:
            self.doModal()

    def onInit(self):
        self.defineControls()
        if not self.parser.feedsList:
            xbmcgui.Dialog().ok(getLS(40) + 'RssFeeds.xml', 'RssFeeds.xml ' + getLS(32041), getLS(32042), getLS(32043))
            self.closeDialog()
        self.showDialog()

    def defineControls(self):
        #actions
        self.action_cancel_dialog = ( 9, 10, 92, 216, 247, 257, 275, 61467, 61448 )
        #control ids
        self.control_heading_label_id = 2
        self.control_list_label_id = 4
        self.control_list_id = 10
        self.control_modifySet_button_id = 11
        self.control_add_button_id = 13
        self.control_remove_button_id = 14
        self.control_ok_button_id = 18
        self.control_cancel_button_id = 19
        #controls
        self.heading_label = self.getControl(self.control_heading_label_id)
        self.list_label = self.getControl(self.control_list_label_id)
        self.list = self.getControl(self.control_list_id)
        self.add_button = self.getControl(self.control_add_button_id)
        self.remove_button = self.getControl(self.control_remove_button_id)
        self.modifySet_button = self.getControl(self.control_modifySet_button_id)
        self.ok_button = self.getControl(self.control_ok_button_id)
        self.cancel_button = self.getControl(self.control_cancel_button_id)
        #defaults
        self.dFeedsList = [{'url':'http://feeds.feedburner.com/xbmc', 'updateinterval':'30'},
                           {'url':'http://feeds.feedburner.com/latest_xbmc_addons', 'updateinterval':'30'},
                           {'url':'http://feeds.feedburner.com/updated_xbmc_addons', 'updateinterval':'30'}]

    def showDialog(self):
        self.heading_label.setLabel(getLS(32030))
        self.list_label.setLabel(getLS(32024))
        self.modifySet_button.setLabel(getLS(32006))
        self.updateSetsList()
        self.setFocus(self.list)

    def closeDialog(self):
        """Close the Set Editor Dialog and open RSS Editor Dialog"""
        from . import rssEditor
        rssEditorUI = rssEditor.GUI("script-RSS_Editor.xml", CWD, "default", setNum = self.setNum)
        self.close()
        del rssEditorUI

    def onClick(self, controlId):
        #select existing set
        if controlId == self.control_list_id:
            setItem = self.list.getSelectedItem()
            self.setNum = setItem.getLabel()
            self.parser.writeXmlToFile()
            self.closeDialog()
        #add new set
        elif controlId == self.control_add_button_id:
            self.getNewSet()
            self.updateSetsList()
        #remove existing set
        elif controlId == self.control_remove_button_id:
            self.removeSet()
            self.updateSetsList()
        #modify existing set
        elif controlId == self.control_modifySet_button_id:
            self.editSet()
            self.updateSetsList()
        #write sets to file/dialog to modify feeds within set.
        elif controlId == self.control_ok_button_id:
            self.parser.writeXmlToFile()
            self.closeDialog()
        #cancel dialog
        elif controlId == self.control_cancel_button_id:
            self.closeDialog()

    def onAction(self, action):
        if action in self.action_cancel_dialog:
            self.closeDialog()

    def onFocus(self, controlId):
        pass

    def editSet(self):
        """Edit the attributes of an existing set"""
        setItem = self.list.getSelectedItem()
        oldSetLabel = setItem.getLabel()
        #ask user for set number
        newSetNum = self.getSetNum(oldSetLabel[3:])
        if newSetNum:
            newSetLabel = 'set' + newSetNum
            #ask user if set contains right to left text
            rtl = self.containsRTLText()
            #copy settings from old label
            self.parser.feedsList[newSetLabel] = self.parser.feedsList[oldSetLabel]
            #apply new attributes
            self.parser.feedsList[newSetLabel]['attrs'] = {'rtl':rtl, 'id':newSetNum}
            #if the set# changes, remove the old one.
            if newSetLabel != oldSetLabel:
                self.removeSet(oldSetLabel)

    def getNewSet(self):
        """Add a new set with some default values"""
        #default setNumber = find highest numbered set, then add 1
        defaultSetNum = max([int(setNum[3:]) for setNum in self.parser.feedsList.keys()]) + 1
        #ask user for set number
        newSetNum = self.getSetNum(defaultSetNum)
        #check if set number already exists
        if newSetNum:
            newSetLabel = 'set' + newSetNum
            #ask user if set contains right to left text
            rtl = self.containsRTLText()
            #add default information
            self.parser.feedsList[newSetLabel] = {'feedslist':self.dFeedsList, 'attrs':{'rtl':rtl, 'id':newSetNum}}

    def getSetNum(self, defaultSetNum, title = getLS(32025)):
        newSetNum = str(xbmcgui.Dialog().numeric(0, title, str(defaultSetNum)))
        if self.setNumExists(newSetNum) and newSetNum != defaultSetNum:
            self.getSetNum(defaultSetNum, getLS(32050) % newSetNum)
        else:
            return newSetNum

    def setNumExists(self, setNum):
        if 'set' + setNum in self.parser.feedsList.keys():
            return True

    def containsRTLText(self):
        """Returns xml style lowercase 'true' or 'false'"""
        return str(bool(xbmcgui.Dialog().yesno(getLS(32027), getLS(32027)))).lower()

    def removeSet(self, setNum = None):
        """Removes a set or if set is required resets it to default"""
        if setNum is None:
            setNum = self.list.getSelectedItem().getLabel()
        if setNum == 'set1':
            #Ask if user wants to set everything to default.
            if xbmcgui.Dialog().yesno(getLS(32045), getLS(32046) + ' ' + getLS(32047)):
                self.parser.feedsList[setNum] = {'feedslist':self.dFeedsList, 'attrs':{'rtl':'false','id':'1'}}
        else:
            del self.parser.feedsList[setNum]

    def updateSetsList(self):
        self.list.reset()
        for setNum in sorted(self.parser.feedsList.keys()):
            self.list.addItem(setNum)
            self.list_label.setLabel(getLS(32024))
